sparse_autoencoder/train/metrics/metric_class.py

```diff
@@ -0,0 +1,30 @@
+"""Base class for metrics."""
+from abc import ABC, abstractmethod
+from typing import TypedDict
+
+from jaxtyping import Float
+from torch import Tensor
+from torch.optim import Optimizer
+
+from sparse_autoencoder.autoencoder.model import SparseAutoencoder
+
+
+class MetricArgs(TypedDict):
+    """Class to hold arguments to metrics, contains everything needed to compute any metric."""
+
+    step: int
+    batch: Float[Tensor, "batch input_activations"]
+    reconstruction_loss_mse: Float[Tensor, " item"]
+    l1_loss: Float[Tensor, " item"]
+    autoencoder: SparseAutoencoder
+    optimizer: Optimizer
+    learned_activations: Float[Tensor, "batch learned_activations"]
+    reconstructed_activations: Float[Tensor, "batch input_activations"]
+
+
+class Metric(ABC):
+    """Base class for metrics."""
+
+    @abstractmethod
+    def compute_and_log(self, args: MetricArgs) -> None:
+        """Compute the metric value."""
```

Comment:
Definitely think it makes sense to have a base Metric class - but I imagine there are 3 types of abstract class that extend from this - metrics during data generation, metrics during training and validation metrics (e.g. total loss on the source model with and without the sae patched in)

Reply:
Should also flag that we ideally don't want the metrics class to store state, unless absolutely necessary. Where possible we can just log things to WandB and then calculate metrics that track history there